File: Others/047.Permutations-II/047.Permutations-II.py
class Solution:
    def permuteUnique(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        def insert(P,k):
            result = []
            for p in P:
                for i in range(len(p)+1):                    
                    result.append(p[:i]+[k]+p[i:])
                    if i<len(p) and p[i]==k: break
            return result
        
        P = [[]]
        for k in nums:
            P = insert(P,k)
        return P

File: Others/047.Permutations-II/047.Permutations-II.cpp
class Solution {
public:
    vector<vector<int>> results;
    vector<vector<int>> permuteUnique(vector<int>& nums) 
    {
        vector<int>perm;
        DFS(perm,0,nums);
        return results;
    }
    
    void DFS(vector<int>perm, int k, vector<int>& nums)
    {
        if (k==nums.size())
        {
            results.push_back(perm);
            return;
        }
        
        for (int i=0; i<k+1; i++)
        {
            vector<int>newPerm = perm;
            if (i==0 || i>=1 && nums[k]!=perm[i-1])
            {
                newPerm.insert(newPerm.begin()+i, nums[k]);
                DFS(newPerm, k+1, nums);
            }
            else
                return;            
        }
            
    }
};
